neqo-crypto/tests/agent.rs

```diff
@@ -530,3 +533,183 @@ fn ech_retry() {
+    impl CertificateCompressor for DefaultEncoding {
+        const ID: u16 = 0x4;
+        const NAME: &CStr = c"copy";
+        const ENABLE_ENCODING: bool = true;
+
+        fn decode(input: &[u8], output: &mut [u8]) -> Res<usize> {
+            let len = std::cmp::min(input.len(), output.len());
+            output[..len].copy_from_slice(&input[..len]);
+            Ok(len)
+        }
+    }
+
+    fixture_init();
+    let mut client = Client::new("server.example", true).expect("should create client");
+    let mut server = Server::new(&["key"]).expect("should create server");
+
+    client
+        .set_certificate_compression::<DefaultEncoding>()
+        .unwrap();
+    server
+        .set_certificate_compression::<DefaultEncoding>()
+        .unwrap();
+
+    connect(&mut client, &mut server);
+
+    assert!(client.state().is_connected());
+    assert!(server.state().is_connected());
+}
+
+struct CopyCompressionNoEncoder {}
+
+impl CertificateCompressor for CopyCompressionNoEncoder {
+    const ID: u16 = 0x4;
+    const NAME: &CStr = c"copy";
+
+    fn decode(input: &[u8], output: &mut [u8]) -> Res<usize> {
+        let len = std::cmp::min(input.len(), output.len());
+        output[..len].copy_from_slice(&input[..len]);
+        Ok(len)
+    }
+}
+
+#[test]
+fn connection_succeeds_when_server_and_client_support_cert_compr_copy_without_encoder() {
+    fixture_init();
+    let mut client = Client::new("server.example", true).expect("should create client");
+    let mut server = Server::new(&["key"]).expect("should create server");
+
+    client
+        .set_certificate_compression::<CopyCompressionNoEncoder>()
+        .unwrap();
+    server
+        .set_certificate_compression::<CopyCompressionNoEncoder>()
+        .unwrap();
+
+    connect(&mut client, &mut server);
+
+    assert!(client.state().is_connected());
+    assert!(server.state().is_connected());
+}
+
+#[test]
+fn connection_succeeds_when_only_server_support_cert_compr() {
+    fixture_init();
+    let mut client = Client::new("server.example", true).expect("should create client");
+    let mut server = Server::new(&["key"]).expect("should create server");
+
+    server
+        .set_certificate_compression::<CopyCompressionNoEncoder>()
+        .unwrap();
+
+    connect(&mut client, &mut server);
+
+    assert!(client.state().is_connected());
+    assert!(server.state().is_connected());
+}
+
+#[test]
+fn connection_succeeds_when_only_client_support_cert_compr() {
+    fixture_init();
+    let mut client = Client::new("server.example", true).expect("should create client");
+    let mut server = Server::new(&["key"]).expect("should create server");
+
+    client
+        .set_certificate_compression::<CopyCompressionNoEncoder>()
+        .unwrap();
+
+    connect(&mut client, &mut server);
+
+    assert!(client.state().is_connected());
+    assert!(server.state().is_connected());
+}
+
+struct CopyCompressionNoEncoderReturnsErr {}
```

Comment:
This type is one that the DECODER returns an error.

You should add a test for when the ENCODER fails.